File: competitiveprogramming/chapter2/A_builtin_libraries/E_queue/A_a_node_too_far/Main.java
import java.util.*;

class Main {
	public static void main(String... args) {
		try {
			Scanner sc = new Scanner(System.in);
			while (true) {
				int N = sc.nextInt();
				if (N == 0) break;
				ArrayList<Node> nodes = new ArrayList<Node>();
				ArrayList<Connection> conns = new ArrayList<Connection>();

				for (int i = 0; i < N; i++) {
					Node a = new Node(sc.nextInt());
					Node b = new Node(sc.nextInt());
					if (index(nodes, a.num) == -1) nodes.add(a);
					if (index(nodes, b.num) == -1) nodes.add(b);
					conns.add(new Connection(a, b));
				}

				while (true) {
					int s = sc.nextInt(), ttl = sc.nextInt();
					if (s == 0 && ttl == 0) break;
					int ans = getAns(nodes, conns, s, ttl);
					System.out.println(String.format(
						"Case %d: %d nodes not reachable from node %d with TTL = %d.",
						1,
						ans,
						s,
						ttl
					));
					return;
				}
			}
		} catch (Exception e) {
			e.printStackTrace();
		} finally {
			System.exit(0);
		}
	}

	static int index(ArrayList<Node> nodes, int s) {
		for (int i = 0; i < nodes.size(); i++)
			if (nodes.get(i).num == s) return i;
		return -1;
	}

	static int getAns(ArrayList<Node> nodes, ArrayList<Connection> conns, int s, int ttl) {
		Node begin = nodes.get(index(nodes, s));
		begin.ttl = ttl;
		begin.mark = true;

		Queue<Node> q = new LinkedList<Node>();
		q.add(begin);

		while (!q.isEmpty()) {
			Node v = q.poll();
			System.out.println("v.num="+v.num+"v.mark="+v.mark);
			if (v.ttl < 0) continue;
			for (int i = 0; i < conns.size(); i++) {
				Connection conn = conns.get(i);

				if (conn.a.num == v.num && !conn.b.mark) {
					System.out.println("Adding conn "+v.num+" "+conn.b.num);
					System.out.println("conn.b.num"+conn.b.num+",conn.b.mark"+conn.b.mark);
					conn.b.mark = true;
					conn.b.ttl  = v.ttl - 1;
					q.add(conn.b);
				}

				if (conn.b.num == v.num && !conn.a.mark) {
					System.out.println("Adding conn "+v.num+" "+conn.a.num);
					System.out.println("conn.a.num"+conn.a.num+",conn.a.mark"+conn.a.mark);
					conn.a.mark = true;
					conn.a.ttl = v.ttl - 1;
					q.add(conn.a);
				}
			}
		}

		int ans = 0;
		for (int i = 0; i < nodes.size(); i++) {
			Node node = nodes.get(i);
			System.out.println("ttl:" + node.ttl + " " + ttl);
			ans += (node.ttl > ttl) ? 1 : 0;
			//node.ttl = -1;
			//node.mark = false;
		}
		return ans;
	}

	static class Node {
		int num, ttl;
		boolean mark;

		public Node(int num) {
			this.num = num;
			this.ttl = -1;
			this.mark = false;
		}
	}

	static class Connection {
		Node a, b;

		public Connection(Node a, Node b) {
			this.a = a;
			this.b = b;
		}
	}

}

File: competitiveprogramming/chapter2/A_builtin_libraries/E_queue/A_a_node_too_far/Main.c
#include <stdio.h>
#include <stdlib.h>

#define LIM 1000

typedef struct {
	int a, b;
} Conn;

int indexOf(int *arr, int n, int itm);
int getAns(Conn *conns, int N, int s, int ttl, int cnt, int *nodes, int *marks);

int main() {
	int caseNum = 1;
	while (1) {
		int N, cnt = 0, i;
		scanf("%d", &N);
		if (N == 0) break;
		Conn *conns = malloc(N * sizeof(Conn));
		int *nodes = malloc(LIM * sizeof(int));
		for (i = 0; i < N; i++) {
			int a, b;
			scanf("%d %d", &a, &b);
			conns[i].a = a; conns[i].b = b;
			if (indexOf(nodes, LIM, a) == -1)
				nodes[cnt++] = a;
			if (indexOf(nodes, LIM, b) == -1)
				nodes[cnt++] = b;
		}
		while (1) {
			int s, ttl;
			scanf("%d %d", &s, &ttl);
			if (s == 0 && ttl == 0) break;
			int *marks = malloc(cnt * sizeof(int));
			for (i = 0; i < cnt; i++) marks[i] = 0;
			int ans = getAns(conns, N, s, ttl, cnt, nodes, marks);
			printf(
				"Case %d: %d nodes not reachable from node %d with TTL = %d.\n",
				caseNum++,
				cnt - ans,
				s,
				ttl
			);
			free(marks);
		}
		free(nodes);
		free(conns);
	}
	return 0;
}

int indexOf(int *arr, int n, int itm) {
	int i;
	for (i = 0; i < n; i++) {
		if (arr[i] == itm) {
			return i;
		}
	}
	return -1;
}

int getAns(Conn *conns, int N, int s, int ttl, int cnt, int *nodes, int *marks) {
	marks[indexOf(nodes, cnt, s)] = 1;
	if (ttl == 0) return 1;
	int i, ans = 1;
	for (i = 0; i < N; i++) {
		if (conns[i].a == s && marks[indexOf(nodes, cnt, conns[i].b)] == 0) {
			ans += getAns(conns, N, conns[i].b, ttl-1, cnt, nodes, marks);
		}
		if (conns[i].b == s && marks[indexOf(nodes, cnt, conns[i].a)] == 0) {
			ans += getAns(conns, N, conns[i].a, ttl-1, cnt, nodes, marks);
		}
	}
	return ans;
}